vaporetto/src/predictor.rs

```diff
@@ -1,1136 +1,861 @@
+}
 
-    /// Predicts word boundaries. This function inserts scores.
-    ///
-    /// # Arguments
-    ///
-    /// * `sentence` - A sentence.
-    ///
-    /// # Returns
-    ///
-    /// A sentence with predicted boundary information.
-    pub fn predict_with_score(&self, sentence: Sentence) -> Sentence {
-        let mut sentence = self.predict_impl(sentence);
-        sentence
-            .boundary_scores
-            .rotate_left(self.data.padding.into());
-        sentence.boundary_scores.truncate(sentence.boundaries.len());
-        sentence
+#[cfg(feature = "tag-prediction")]
+impl AddAssign<&Self> for PositionalWeightWithTag {
+    fn add_assign(&mut self, other: &Self) {
+        if let Some(y) = self.weight.as_mut() {
+            if let Some(x) = other.weight.as_ref() {
+                *y += x;
+            }
+        } else {
+            self.weight = other.weight.clone();
+        }
+        for (k, v) in &other.tag_info {
+            self.tag_info
+                .entry(*k)
+                .and_modify(|w| {
+                    for (y, x) in w.iter_mut().zip(v) {
+                        *y += *x;
+                    }
+                })
+                .or_insert_with(|| v.clone());
+        }
     }
+}
+
+#[cfg(feature = "tag-prediction")]
+#[derive(Decode, Encode)]
+struct TagPredictor {
+    tags: Vec<Vec<String>>,
+    bias: WeightVector,
+}
+
+#[cfg(feature = "tag-prediction")]
+impl TagPredictor {
+    pub fn new(tags: Vec<Vec<String>>, bias: Vec<i32>) -> Self {
+        Self {
+            tags,
+            bias: bias.into(),
+        }
+    }
+
+    #[inline]
+    pub const fn bias(&self) -> &WeightVector {
+        &self.bias
+    }
+
+    #[inline]
+    pub fn predict<'a>(&'a self, scores: &[i32], tags: &mut [Option<Cow<'a, str>>]) {
+        let mut offset = 0;
+        for (tag_cands, tag) in self.tags.iter().zip(tags) {
+            if tag_cands.len() >= 2 {
+                let mut idx = 0;
+                let mut max_score = i32::MIN;
+                for (i, &s) in scores[offset..offset + tag_cands.len()].iter().enumerate() {
+                    if s > max_score {
+                        idx = i;
+                        max_score = s;
+                    }
+                }
+                tag.replace(Cow::Borrowed(&tag_cands[idx]));
+                offset += tag_cands.len();
+            } else {
+                *tag = tag_cands.first().map(|t| Cow::Borrowed(t.as_str()));
+            }
+        }
+    }
+}
+
+pub struct PredictorData {
+    char_scorer: Option<CharScorer>,
+    type_scorer: Option<TypeScorer>,
+    bias: i32,
 
     #[cfg(feature = "tag-prediction")]
-    fn best_tag(&self, scores: &[i32]) -> Arc<String> {
-        Arc::clone(
-            scores
-                .iter()
-                .zip(&self.data.tag_names)
-                .max_by_key(|(&x, _)| x)
-                .unwrap()
-                .1,
-        )
+    tag_predictor: HashMap<String, (u32, TagPredictor)>,
```

Comment:
`predict_tags` can perform even if `new` takes `predict_tags=false`, but it is not expected.
This could be solved by using `Option` to this member.

Reply:
d2acb55